Correct missing sizes for manifest schema 1 images

miminar opened the pull request:
Instead of filling metadata in the image object in the registry, send the manifest and config blobs to the master and let it do the job.

Resolves #16306
Resolves: [bz#1491589](https://bugzilla.redhat.com/show_bug.cgi?id=1491589)

miminar commented:
@dmage PTAL
@mfojtik FYI

miminar commented:
The code is covered by `extended_image_registry` tests which will be part of a nightly job.

miminar commented:
/retest

dmage commented on pkg/dockerregistry/server/manifestschema1handler.go:
```diff
@@ -102,7 +102,7 @@ func (h *manifestSchema1Handler) FillImageMetadata(ctx context.Context, image *i
 	image.DockerImageMetadata.Object = meta
 
-	return nil
+	return encodeRawDockerImageMetadata(image, true)
```
above, imageMetadataFromManifest is called, which calls encodeRawDockerImageMetadata. Can't we merge the code above into imageMetadataFromManifest?

kargakis commented:
/retest

smarterclayton commented:
Add test so this doesn't regress

miminar commented:
I moved the metadata filling into separate schema handlers since they differ a lot. I've also added few unit tests. I'll add some more coverage, especially for schema2. 

@legionus, @dmage PTAL

dmage commented on pkg/dockerregistry/server/manifestschema1handler.go:
```diff
@@ -53,56 +53,75 @@ type manifestSchema1Handler struct {
 var _ ManifestHandler = &manifestSchema1Handler{}
 
 func (h *manifestSchema1Handler) FillImageMetadata(ctx context.Context, image *imageapiv1.Image) error {
```
As it's our internal type, I'd like to propose to split this function into 3 parts:

    Layers(ctx context.Context) (layers []imageapi.ImageLayer, annotations map[string]string, error)
    Signatures(ctx context.Context) ([][]byte, error)
    Metadata(ctx context.Context, layers []imageapi.ImageLayer) (imageapi.DockerImage, err error)

The output of these functions should depend only on the input manifest. And we do not need to parse manifest data in this functions, because we already have `h.manifest` (in this case we can remove DockerImageManifest in caller).

These functions can be reused in pkg/image/apis/image.ImageWithMetadata.

miminar replied:
Good idea, I'll rework.

miminar replied:
Reworked. I modified the declarations a bit. PTAL

smarterclayton commented:
sgtm

On Mon, Sep 18, 2017 at 3:41 PM, Oleg Bulatov <notifications@github.com>
wrote:

> *@dmage* commented on this pull request.
> ------------------------------
>
> In pkg/dockerregistry/server/manifestschema1handler.go
> <https://github.com/openshift/origin/pull/16325#discussion_r139519999>:
>
> > @@ -53,56 +53,75 @@ type manifestSchema1Handler struct {
>  var _ ManifestHandler = &manifestSchema1Handler{}
>
>  func (h *manifestSchema1Handler) FillImageMetadata(ctx context.Context, image *imageapiv1.Image) error {
>
> As it's our internal type, I'd like to propose to split this function into
> 3 parts:
>
> Layers(ctx context.Context) (layers []imageapi.ImageLayer, annotations map[string]string, error)
> Signatures(ctx context.Context) ([][]byte, error)
> Metadata(ctx context.Context, layers []imageapi.ImageLayer) (imageapi.DockerImage, err error)
>
> The output of these functions should depend only on the input manifest.
> And we do not need to parse manifest data in this functions, because we
> already have h.manifest (in this case we can remove DockerImageManifest
> in caller).
>
> These functions can be reused in pkg/image/apis/image.ImageWithMetadata.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/16325#pullrequestreview-63464116>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pyWLOuxLD_ZKo0iF9SbhzyrTo5X1ks5sjsdQgaJpZM4PWPJK>
> .
>

miminar commented:
Added more tests.
Instead of filling metadata to the image object in the registry, the manifest and config blobs are sent to master API, which should do the job.

miminar commented:
/test extended_image_regstry

dmage commented on pkg/dockerregistry/server/manifestservice.go:
```diff
@@ -143,6 +148,7 @@ func (m *manifestService) Put(ctx context.Context, manifest distribution.Manifes
 			DockerImageReference:         fmt.Sprintf("%s/%s/%s@%s", m.repo.config.registryAddr, m.repo.namespace, m.repo.name, dgst.String()),
 			DockerImageManifest:          string(payload),
 			DockerImageManifestMediaType: mediaType,
+			DockerImageConfig:            string(config),
```
It will land in etcd, won't it?

miminar replied:
> It will land in etcd, won't it?

I've updated the image strategy to unset `DockerImageConfig` and `DockerImageManifest`.

miminar commented:
/test extended_image_registry

miminar commented:
The annotation `openshift.io/image.manifestBlobLocations=registry` is a protection against removing manifest/config from image object when image is created/updated outside of latest registry version that makes sure to store the blobs on its storage and setting the annotation.

dmage commented on pkg/image/apis/image/types.go:
```diff
@@ -51,6 +51,14 @@ const (
 
+	// DockerImageManifestBlobLocationsAnnotation is a comma-separated list with locations of image's manifest
+	// blob and its configuration blob.
+	DockerImageManifestBlobLocationsAnnotation = "openshift.io/image.manifestBlobLocations"
```
Do you expect values other than "registry"?

miminar replied:
There are three different possible locations: `registry`, `etcd` and `external`. We may want to record known manifest locations in the future (like docker daemon). The annotation would allow for encoded urls as well. Although, some new image attribute would fit better for such a purpose.

smarterclayton replied:
If you plan on using URLs in the future you should be using an array of structs, JSON encoded.

smarterclayton replied:
Also, constants used in apis should always be `CamelCase`

miminar replied:
> If you plan on using URLs in the future you should be using an array of structs, JSON encoded.

I'd rather avoid it using annotation. Let's add another attribute if the need arises in the future.

I've turned the annotation into boolean `"image.openshift.io/manifestBlobLocations"`.

> Also, constants used in apis should always be CamelCase

I may be missing something, but all the other image annotation names are in `lowerCamelCase`. On the other hand, builds [use](https://github.com/openshift/origin/blob/848776c00ea084f991a6a8b51c8f20fecc53efbc/pkg/build/apis/build/types.go#L15) `dash-separated-lower-case`.

miminar commented:
Another rebase needed.

miminar commented:
/test extended_image_registry

miminar commented:
@stevekuznetsov 

When I click on detail link of ci/openshift-jenkins/extended_image_registry, I get:

```
Unable to load build details from gs://origin-ci-test/pr-logs/pull/16325/test_pull_request_origin_extended_image_registry/24
This may mean the job has not yet uploaded any artifacts, or that it failed to upload. Take a look at the job in Jenkins for more information.
```

The job link works fine though. Is this openshift/test-infra issue?

miminar commented:
/retest

stevekuznetsov commented:
@miminar yeah we had some issues yesterday... should be fixed from now on :)

miminar commented:
@smarterclayton @dmage any other comments?

smarterclayton commented:
Looks fine to me but I'll let @dmage have the final say

bparees commented:
@dmage can you sign off on this?  looks like this is the fix for a blocking bug.

bparees commented:
(it also appears to be blocking https://github.com/openshift/origin/pull/15807, it would be great to clear this logjam)

dmage commented:
/lgtm

legionus commented:
@miminar @dmage Should we add the `imageapi.ImageManifestBlobStoredAnnotation` annotation in the `TagService` ?

https://github.com/miminar/origin/blob/5ac2762a6e0874359683a32567350917f7321b19/pkg/dockerregistry/server/tagservice.go#L170-L177

legionus commented:
> Should we add the imageapi.ImageManifestBlobStoredAnnotation annotation in the TagService ?

I will answer myself. No. This should not be done because the image we get from the API.

/lgtm
/approve

miminar commented:
/assign @bparees

bparees commented:
/approve
